General Clean-Up PR

JDBraun opened the pull request:
**General Updates:**

- **Provider Upgrade:** The Databricks provider has been upgraded to 1.46. No impact was seen in usage from 1.46 to the previous version. However, if you're managing an active environment, you should do your own testing in your lower environments.

- **README:** Moved the readme to the top level of the AWS repository for easier access, isolating the relevant Terraform code. Cleaned-up the **Post Workspace Deployment** to only include modules that are deployed by default. All optional modes are outlined in **Optional Deployment Configuration.**

- **Metastore Creation:**
    - Metastore ID variable has been removed. Now it is a boolean value if metastore exists or not in the region. If a metastore exists for that region, then the new workspace will be attached. This will have an impact to the very small percentage of accounts that have multiple metastores in the same region. They can directly add the metastore ID in the metastore assignment TF file. This addresses this issue #67 

- **Variable Updates:** 
    - **Booleans in variables names:** Any optional variable that is true or false now has the same naming convention. Like enable_read_only_external_location_boolean, enable_admin_configs_boolean, etc.
    - **dbfsname variable:** Removed DBFS name as variable to insert. This is now created based on the resource prefix. It is now renamed as "root_s3_bucket" throughout the codebase for additional clarity.
    - **Workspace admin service principal variable:** This is now includes _admin_ in the name to represent to permission level of this created SP.
    - **Compliance security profile egress ports:** This has been renamed to better describe it's functionality now that a compliance security profile resource exists. This enablement only impacts the outbound ports on the workspace SG and inbound ports on the PrivateLink.
    - **Error message on operation_mode:** Updated the error message so it no longer says network type.

- **template.tfvars.example:**
    - The template file now has comments on what values to provide for each of the values for clarity.

**Individual Files:**

- **sra.tf:**
   - The main file has been cleaned up and organized based on required variables, operation mode, and the optional modules that can be enabled on a deployment.
   - Made the default amount of subnets to three instead of two to increase the high availability of the workspace, which is more applicable to production use cases. 

- **networking.tf:**
   - VPC has been renamed from 'data plane' to 'classic compute plane' for clarity
   - Workspace security group has been renamed from 'data plane security group' to 'workspace SG' for clarity

- **privatelink.tf:**
    - Fixed an issue around PrivateLink endpoints not being applied to all availability zones if more than two subnets were provided in the variables.

- **cmk.tf:**
    - Added tags for both the workspace storage and managed key.

- **credential.tf**
   - Removed redundant allow block image resource with the condition key of the Databricks account ID. This is covered by the explicit denied that was added in a [previous commit](https://github.com/databricks/terraform-databricks-sra/commit/11c47190193b5cc1d52746cbebb83ea5b3a2dd92).

- **workspace.tf**
    - Added a commented out deployment_name in the workspace resource configuration. This is not a feature available to all accounts, so this can be commented out and added as a variable for applicable customers.

metrocavich commented on aws/README.md:
```diff
@@ -94,28 +101,33 @@ In this section, we break down additional security recommendations and opportuni
 3. Decide which [operation](https://github.com/databricks/terraform-databricks-sra/tree/main/aws/tf#operation-mode) mode you'd like to use.
 4. Fill out `sra.tf` in place
-5. Fill out `example.tfvars` and place in `tf` directory
+5. Fill out `example.tfvars` and place in `tf` directory, remove the .example part of the file name
```
Isn't the filename `template.tfvars`

JDBraun replied:
yes good catch! will update

metrocavich commented on aws/tf/modules/sra/cmk.tf:
```diff
@@ -58,13 +59,17 @@ resource "aws_kms_key" "workspace_storage" {
+
 resource "aws_kms_alias" "workspace_storage_key_alias" {
-  name          = "alias/${var.resource_prefix}-workspace-storage-key-alias"
+  name          = "alias/${var.resource_prefix}-workspace-storage-key"
```
I love this change. I personally dislike redundant names

JDBraun commented on aws/tf/modules/sra/credential.tf:
```diff
@@ -77,21 +77,6 @@ resource "aws_iam_role_policy" "cross_account" {
       },
-      {
-        "Sid" : "AllowEc2RunInstanceImagePerTag",
-        "Effect" : "Allow",
```
I don't believe so, because we're doing an explicit deny if not resource here https://github.com/JDBraun/terraform-databricks-sra/blob/d6fa97aa553b3209c911497f68f8dcf363dcd56a/aws/tf/modules/sra/credential.tf#L108

I haven't had any issues spinning warehouses or cluster without it

metrocavich commented on aws/tf/modules/sra/network.tf:
```diff
@@ -7,7 +7,7 @@ module "vpc" {
   count = var.operation_mode != "custom" ? 1 : 0
 
-  name = "${var.resource_prefix}-data-plane-VPC"
+  name = "${var.resource_prefix}-classic-compute-plane-vpc"
```
It feels strange to rename one part of networking from data-plane -> classic-compute-plane, but to rename the rest as data-plane -> workspace

Just seems odd to go from 1 name for the data plane to referring to it both as a classic computer plane and also a "workspace"

metrocavich replied:
Super minor comment, just thinking out loud.

JDBraun replied:
no it's fair - the reason I renamed the data plane SG to 'workspace' is that the privatelink sg is 'privatelink', so those are technically both part of the classic compute plane, I thought I'd add another level of granularity for them

metrocavich replied:
Makes a ton of sense to me

metrocavich commented on aws/tf/sra.tf:
```diff
@@ -1,89 +1,77 @@
+
+  // AWS Specific Variables:
+  cmk_admin_arn                            = null // CMK admin ARN, defaults to the AWS account root user.
+  vpc_cidr_range                           = "10.0.0.0/18"
```
Since this file is filled with comments, I can say that I have had confusion when I send over VPC CIDR prepopulated, might be worth adding a comment here just as //Example CIDR, must update private_subnets_cidr, public_subnets_cidr, privatelink_subnets_cidr, and/or firewall_subnets_cidr if changed

JDBraun replied:
like the VPC CIDR is 172.* then the subnets remain the same?

metrocavich replied:
Yeah, the following CIDR ranges would be invalid. Just something like // If updated, must update below subnets

JDBraun commented:
@metrocavich updated the dbfs_s3.tf -> root_s3_bucket.tf, I added the account ID to the bucket policy as per best practices for that, and then I updated the readme

ryan-gord-db commented on aws/tf/modules/sra/variables.tf:
```diff
@@ -168,30 +163,29 @@ variable "ip_addresses" {
   validation {
     condition     = contains(["sandbox", "firewall", "custom", "isolated"], var.operation_mode)
-    error_message = "Invalid network type. Allowed values are: sandbox, firewall, custom, isolated."
+    error_message = "Invalid operation model. Allowed values are: sandbox, firewall, custom, isolated."
```
Minor nit:  `s/model/mode/`

ryan-gord-db reviewed (commented):
`var.resource_owner` is removed which is fine.  However, lots of companies tend to require some type of tag on AWS resources like _billing-code_, _department-code_, etc.  Is it worth adding a _tags_ variable where user-provided tags are applied to the AWS provider's `default_tags` property?

JDBraun commented on aws/tf/modules/sra/databricks_account/uc_assignment/uc_assignment.tf:
```diff
@@ -1,6 +1,11 @@
 // Metastore Assignment
+
+data "databricks_metastore" "this" {
```
Yeah I thought about this on implementation - since this isn't a normal situation, I thought of leaving it and making a comment about it - since I did remove the metastore ID as a variable in general

ryan-gord-db replied:
Fair point.  I imagine the number of folks encountering this scenario to be quite small.  A comment would be a nice addition though methinks.

JDBraun commented:
@metrocavich - updated the sra.tf with a note about the VPC CIDR range

@ryan-gord-db - updated the variables.tf with the right operation mode, then updated the sra.tf with a note on the metastore ID

JDBraun commented:
@metrocavich / @ryan-gord-db - let me know if y'all are good on this PR, or if there's anything else
